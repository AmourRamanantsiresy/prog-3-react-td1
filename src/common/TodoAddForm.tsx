import { ChangeEvent, FormEvent, useState } from 'react';
import { TI_INPUT_TODO_ADD } from '../tests/utils/selectors';
import TodoItem, { TodoStatus } from '../utils/models/Todo';

type TodoAddFormProps = {
  add: (todo: TodoItem) => void;
};

const TodoAddForm = (props: TodoAddFormProps) => {
  const [todoValue, setTodoValue] = useState<string>('');
  const { add } = props;

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if (todoValue.length > 0) {
      add({ id: new Date().getTime(), status: TodoStatus.TODO, value: todoValue });
      setTodoValue('');
    }
  };

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setTodoValue(event.target.value);
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        data-testid={TI_INPUT_TODO_ADD}
        placeholder='Add task'
        value={todoValue}
        type='text'
        onChange={handleChange}
      />
    </form>
  );
};

export default TodoAddForm;
