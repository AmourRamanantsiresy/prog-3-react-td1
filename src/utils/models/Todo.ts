export enum TodoStatus {
  TODO = 'todo',
  DONE = 'done',
}

type Todo = {
  value: string;
  id: number;
  status: TodoStatus;
};

export default Todo;
