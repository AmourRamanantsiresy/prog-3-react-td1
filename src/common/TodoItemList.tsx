import { ReactNode } from 'react';
import Todo, { TodoStatus } from '../utils/models/Todo';
import TodoItem from './TodoItem';

type TodoItemListProps = {
  onDone?: (todoId: number) => void;
  onRemove?: (todoId: number) => void;
  todoStatus: TodoStatus;
  todos: Todo[];
  actions?: ReactNode;
};

const TodoItemList = (props: TodoItemListProps) => {
  const { todoStatus, todos, onDone, onRemove, actions } = props;
  const currentTodos = todos.filter(todo => todo.status === todoStatus);
  return (
    <div style={{ padding: '.5rem' }}>
      <h1>{todoStatus.toString().toUpperCase()}</h1>
      <div className='actions'>{todoStatus === TodoStatus.TODO && actions}</div>
      <div className='list-container'>
        {currentTodos.length > 0 ? (
          currentTodos.map((todo, index) => <TodoItem onRemove={onRemove} key={'input' + index + todo.id} todo={todo} onDone={onDone} />)
        ) : (
          <p className='message-muted'>No task</p>
        )}
      </div>
    </div>
  );
};

export default TodoItemList;
