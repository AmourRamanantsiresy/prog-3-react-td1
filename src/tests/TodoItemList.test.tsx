import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { create, ReactTestRendererJSON } from 'react-test-renderer';
import TodoAddForm from '../components/TodoAddForm';
import TodoItemList from '../components/TodoItemList';
import Todo, { TodoStatus } from '../utils/models/Todo';
import { TODO_LIST_MOCK } from './utils/mocks/todo-mock';
import { ID_TODO_ADD, ID_TODO_DELETE, ID_TODO_DONE, ROLE_TODO_DELETE } from './utils/selectors';

describe('Test TodoItemList', () => {
  test('Architecture', () => {
    const tree = create(
      <TodoItemList
        key='todo_item_list_1'
        todoStatus={TodoStatus.TODO}
        todos={TODO_LIST_MOCK}
        actions={<TodoAddForm add={todo => console.log(todo)} />}
        onDone={id => console.log(id)}
        onRemove={id => console.log(id)}
      />
    ).toJSON() as ReactTestRendererJSON;

    expect(tree).toMatchSnapshot();
  });

  test('When todo status === TODO', () => {
    const handleDone: jest.Mock<number> = jest.fn(id => id);
    const TODO_TODO_MOCK: Todo[] = TODO_LIST_MOCK.filter(todo => todo.status === TodoStatus.TODO);
    const todoToDone = TODO_TODO_MOCK[0];
    const toDoneSelector = `${ID_TODO_DONE}-${todoToDone.id}`;
    render(
      <TodoItemList
        key='todo_item_list_1'
        todoStatus={TodoStatus.TODO}
        todos={TODO_LIST_MOCK}
        actions={<TodoAddForm add={todo => console.log(todo)} />}
        onDone={handleDone}
        onRemove={id => console.log(id)}
      />
    );

    screen.getByText(TodoStatus.TODO.toString().toLocaleUpperCase());

    expect(screen.queryByText('No task')).toBeNull();
    expect(screen.getByTestId(ID_TODO_ADD)).toBeTruthy();
    expect(screen.queryByRole(ROLE_TODO_DELETE)).toBeNull();
    expect(screen.queryAllByRole('checkbox').length).toBe(TODO_TODO_MOCK.length);

    userEvent.click(screen.getByTestId(toDoneSelector));
    expect(handleDone.mock.results[0].value).toEqual(todoToDone.id);
  });

  test('When todo status === DONE', () => {
    const handleRemove: jest.Mock<number> = jest.fn(id => id);
    const TODO_DONE_MOCK: Todo[] = TODO_LIST_MOCK.filter(todo => todo.status === TodoStatus.DONE);
    const todoToRemove = TODO_DONE_MOCK[0];
    const toRemoveSelector = `${ID_TODO_DELETE}-${todoToRemove.id}`;
    render(
      <TodoItemList
        key='todo_item_list_1'
        todoStatus={TodoStatus.DONE}
        todos={TODO_LIST_MOCK}
        actions={<TodoAddForm add={todo => console.log(todo)} />}
        onDone={id => console.log(id)}
        onRemove={handleRemove}
      />
    );

    screen.getByText(TodoStatus.DONE.toString().toLocaleUpperCase());

    expect(screen.queryByText('No task')).toBeNull();
    expect(screen.queryByRole('checkbox')).toBeNull();
    expect(screen.queryByTestId(ID_TODO_ADD)).toBeNull();
    expect(screen.queryAllByRole(ROLE_TODO_DELETE).length).toBe(TODO_DONE_MOCK.length);

    userEvent.click(screen.getByTestId(toRemoveSelector));
    expect(handleRemove.mock.results[0].value).toEqual(todoToRemove.id);
  });

  test('When there is no todo', () => {
    render(
      <TodoItemList
        key='todo_item_list_1'
        todoStatus={TodoStatus.DONE}
        todos={[]}
        actions={<TodoAddForm add={todo => console.log(todo)} />}
        onDone={id => console.log(id)}
        onRemove={id => console.log(id)}
      />
    );

    screen.getByText(TodoStatus.DONE.toString().toLocaleUpperCase());

    expect(screen.queryByText('No task')).toBeTruthy();
    expect(screen.queryByRole('checkbox')).toBeNull();
    expect(screen.queryByRole(ROLE_TODO_DELETE)).toBeNull();
    expect(screen.queryByTestId(ID_TODO_ADD)).toBeNull();
  });
});
