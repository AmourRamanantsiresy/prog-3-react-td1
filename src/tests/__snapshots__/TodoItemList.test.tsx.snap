// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoItemList Architecture 1`] = `
<div
  style={
    Object {
      "padding": ".5rem",
    }
  }
>
  <h1>
    TODO
  </h1>
  <div
    className="actions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        data-testid="input-todo-add-id"
        onChange={[Function]}
        placeholder="Add task"
        type="text"
        value=""
      />
    </form>
  </div>
  <div
    className="list-container"
  >
    <div
      className="flex todo-item"
    >
      <input
        className="pointer"
        data-testid="input-todo-done-id-1"
        id="1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="ml-2"
      >
        Todo number 1
      </label>
    </div>
    <div
      className="flex todo-item"
    >
      <input
        className="pointer"
        data-testid="input-todo-done-id-3"
        id="3"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="ml-2"
      >
        Todo number 3
      </label>
    </div>
    <div
      className="flex todo-item"
    >
      <input
        className="pointer"
        data-testid="input-todo-done-id-5"
        id="5"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="ml-2"
      >
        Todo number 5
      </label>
    </div>
    <div
      className="flex todo-item"
    >
      <input
        className="pointer"
        data-testid="input-todo-done-id-7"
        id="7"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="ml-2"
      >
        Todo number 7
      </label>
    </div>
    <div
      className="flex todo-item"
    >
      <input
        className="pointer"
        data-testid="input-todo-done-id-9"
        id="9"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="ml-2"
      >
        Todo number 9
      </label>
    </div>
  </div>
</div>
`;
